### PR TITLE
Add PhpUnitSetUpTearDownVisibilityFixer

### PR DESCRIPTION
**Pull request by BackEndTea**
This fixer changes the setup and teardown function of phpunit classes to protected, to match its parent.

As talked about in #2538

This is ready for review

### CODE REVIEW

**Comment by SpacePossum** [comment]
Hi, thanks for the PR, good idea to get this done :)

File: README.rst [hunk 1]
```diff
@@ -1125,6 +1125,11 @@ Choose from the list of available rules:
     ``'newest'``
   - ``use_class_const`` (``bool``): use ::class notation; defaults to ``true``
 
+* **php_unit_set_up_tear_down_visibility**
+
+  Changes the visibility of the setUp and tearDown functions of phpunit to
+  protected, to match its parent.
```
Comment (SpacePossum) [inline README.rst]: I would reword a bit, as the new casing might not be the one from the direct parent, it should always be the one of the PHPUnit class I think?

File: src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        return $tokens->isAllTokenKindsFound([T_CLASS, T_FUNCTION]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach (array_reverse($this->findPhpUnitClasses($tokens)) as $indexes) {
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: Maybe it is more efficient to search backwards and store the result, than the `array_reverse` is not needed.
Even more optimized code is possible by using `yield` in the `findPhpUnitClasses`, but this is not the most important optim here ;)

File: src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    /**
+     * @param Tokens $tokens
+     * @param int    $startIndex
+     * @param int    $endIndex
+     */
+    private function fixSetUpAndTearDown(Tokens $tokens, $startIndex, $endIndex)
+    {
+        $tokensAnalyzer = new TokensAnalyzer($tokens);
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: please keep two flags, when both methods are seen (and fixed or not doesn't matter); do an early return

File: src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    private function isMethodSetUpOrTearDown(Tokens $tokens, $index)
+    {
+        $functionNameIndex = $tokens->getNextMeaningfulToken($index);
+        $functionName = \strtolower($tokens[$functionNameIndex]->getContent());
+
+        if ('setup' !== $functionName && 'teardown' !== $functionName) {
+            return false;
+        }
+
+        return true;
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: can be one line `return 'setup' !== $functionName && 'teardown' !== $functionName;`

File: tests/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+class FixerTest extends \PhpUnit\FrameWork\TestCase
+{
+
+    public function setUp() {}
+
+    public function tearDown() {}
+}
+',
+            ],
+            'setUp and tearDown stay protected if they are' => [
```
Comment (SpacePossum) [inline tests/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixerTest.php]: great idea, however you already test this as part of the `expected` result in the previous test case

File: src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+            $visibility = $tokensAnalyzer->getMethodAttributes($i)['visibility'];
+
+            if (T_PROTECTED === $visibility) {
+                continue;
+            }
+
+            if (T_PUBLIC === $visibility) {
+                do {
+                    $i = $tokens->getPrevMeaningfulToken($i);
+                } while (!$tokens[$i]->isGivenKind(T_PUBLIC));
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: you can use `$tokens->getPrevTokenOfKind` to find the `T_PUBLIC` as replacement of this loop

File: src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    /**
+     * @param Tokens $tokens
+     * @param int    $startIndex
+     * @param int    $endIndex
+     */
+    private function fixSetUpAndTearDown(Tokens $tokens, $startIndex, $endIndex)
+    {
+        $tokensAnalyzer = new TokensAnalyzer($tokens);
+        for ($i = $endIndex - 1; $i > $startIndex; --$i) {
+            if (!$this->isMethod($tokens, $i)) {
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: the `isMethod` method content can be moved into `isMethodSetUpOrTearDown`, but that is just my test

File: src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+                continue;
+            }
+            if (!$this->isMethodSetUpOrTearDown($tokens, $i)) {
+                continue;
+            }
+
+            $visibility = $tokensAnalyzer->getMethodAttributes($i)['visibility'];
+
+            if (T_PROTECTED === $visibility) {
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: `|| T_PRIVATE`
or maybe more defensive would to act only `T_PUBLIC` or `null`

Comment (BackEndTea) [reply src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: No problem, but having a setUp or tearDown as a private function would throw an error on execution of the tests.

Comment (SpacePossum) [reply src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: indeed it would, however we might be fixing a non-PHPUnit class here that happens to look like one

File: src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php (hunk 2 above)
Comment (BackEndTea) [reply src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: I'm not quite sure how to do that, so if possible, i'd like some help with that.

File: tests/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixerTest.php (hunk 5 above)
Comment (BackEndTea) [reply tests/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixerTest.php]: True, but this test explicitly makes sure nothing changes if they are protected.
So even if it may be redundant, i personally like having that extra check. If needed i can remove it though.

**Comment by BackEndTea** [comment]
@SpacePossum This is once again ready for review

File: tests/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixerTest.php (hunk 5 above)
Comment (SpacePossum) [reply tests/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixerTest.php]: I understand, but you already have the extra test, when ever you pass;
```
        $this->doTest(
            $a,
            $b
        );
```

it will always also test;

```
        $this->doTest(
            $a,
            null
        );
```

so if you have two test with same value for `$a` than one repeats the other

File: src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php (hunk 2 above)
Comment (SpacePossum) [reply src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: the easiest way would be to iterate backwards in `findPhpUnitClasses`,

```php
for ($index = $tokens->count() - 1; $index >= 0; --$index) {
```

there should be bunch of fixers that do this already which might help as an example. Maybe I can help you out later with this.

File: tests/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixerTest.php [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+ *
+ * @covers \PhpCsFixer\Fixer\PhpUnit\PhpUnitSetUpTearDownVisibilityFixer
+ */
+final class PhpUnitSetUpTearDownVisibilityFixerTest extends AbstractFixerTestCase
+{
+    /**
+     * @dataProvider provideFixCases
+     *
+     * @param mixed      $expected
+     * @param null|mixed $input
```
Comment (SpacePossum) [inline tests/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixerTest.php]: little left over; `mixed` should be `string` I think

File: src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php [hunk 10]
```diff
@@ -67,31 +67,27 @@ public function isCandidate(Tokens $tokens)
...
-        for ($index = 0, $limit = $tokens->count() - 1; $index < $limit; ++$index) {
+        for ($index = $tokens->count() - 1; $index >= 0; --$index) {
             if ($tokens[$index]->isGivenKind(T_CLASS) && $phpUnitTestCaseIndicator->isPhpUnitClass($tokens, $index)) {
-                $index = $tokens->getNextTokenOfKind($index, ['{']);
-                $endIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $index);
-                $phpunitClasses[] = [$index, $endIndex];
-                $index = $endIndex;
+                $start = $tokens->getNextTokenOfKind($index, ['{']);
+                $endIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $start);
+                yield [$start, $endIndex];
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitSetUpTearDownVisibilityFixer.php]: thanks for this change, code looks nice and fast : )
I'll test again soon

**Comment by SpacePossum** [comment]
@BackEndTea if you've the time, could you rebase your branch and add this new fixer to the [.php_cs.dist](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/.php_cs.dist) of the project?

**Comment by keradus** [comment]
> btw. is this fixer `risky`

wow, why it is not marked as risky ?

**Comment by BackEndTea** [comment]
@keradus @SpacePossum Is this good for the next release, or are there still things that need to be fixed?

**Comment by SpacePossum** [comment]
@BackEndTea please pull my commits in if you want to keep up working in this.
I think this looks good, but we've to wait for people to find the time to merge and release and such, please be patient with us, busy days! :)

**Comment by BackEndTea** [comment]
No problem, just wanted to make sure that you weren't waiting for something from my side.

**Comment by keradus** [comment]
@BackEndTea 
@SpacePossum , please finish review

**Comment by SpacePossum** [comment]
I finished my review 19 days ago, we are waiting for a second OK

**Comment by keradus** [comment]
ah, didn't saw your approval on PR ;)

**Comment by keradus** [comment]
Thank you @BackEndTea.

**Comment by BackEndTea** [comment]
Thank you @keradus @SpacePossum
